### PR TITLE
chore!: Make file manager read-only to the compiler

### PR DESCRIPTION
**Pull request by kevaundray**
# Description

This is a PR that moves us towards having the compiler see the FileManager as an immutable object. How it gets populated will now be upto the caller.

One issue we have so far is that caller, can be in the wasm context or the native context, so we have file_reader being a component that we conditionally compile depending on the context. 

Ideally we move to a case where both package source resolution and dependency source resolution is done beforehand, and the compiler assumes that FileManager has both the source for its current crate and all of the crates dependencies.

This separation of concerns makes it easier to incrementally compile parts of the compiler in the future. For example, if the FileManager has not changed, then we know that the compilation result should not change,and we do not need to recompile.

Further, lets say we add a pass that solely does parsing of the root files in the File Manager, If a particular parsed AST has not changed, then we do not need to parse that file again.  

## Problem\*

Resolves 

## Documentation\*

Check one:
- [ ] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[Exceptional Case]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by kevaundray** [comment]
One difficulty with this PR is how to handle things in the browser context, where there are no files.

I think initially we can support the usecase of a single file in the browser, since this doesn't break existing playgrounds (This allows us to mitigate the file system issue). For node, we do have the file system and we should try to not break that workflow.

File: tooling/nargo/src/lib.rs
Comment (kevaundray) [inline tooling/nargo/src/lib.rs:83]:
```suggestion
// TODO: given that this crate should not know about the file manager.
```

**Comment by kevaundray** [comment]
The remarkable errors are due to how we special handle the stdlib -- we may need to move this logic outside of the file manager (which makes sense since the file manager should only be concerned with files and not need to know about the stdlib

**Comment by kevaundray** [comment]
Last set of errors should be related to the file manager not having the source code for its dependencies (since the method we added only allows us to add code for the current package)

File: compiler/noirc_driver/src/lib.rs
Comment (TomAFrench) [inline compiler/noirc_driver/src/lib.rs:83]:
```suggestion
        let path = Path::new(&path);
        context.file_manager.add_file_with_source(path, source);
```

File: tooling/nargo/src/lib.rs
Comment (TomAFrench) [inline tooling/nargo/src/lib.rs:60]: This can get expensive as it's reading in stuff outside of just the `src` dir so if there's other stuff inside the Nargo project such as a `node_modules` directory, we could be sucking up gigabytes of data here.

We should read the entrypoint from the `Nargo.toml` file and just read from that directory.

Comment (TomAFrench) [reply tooling/nargo/src/lib.rs:60]: Not saying it needs to be handled right now, just making a note.

Comment (kevaundray) [reply tooling/nargo/src/lib.rs:60]: yes sir, will fixup once the wasm codepath is working

**Comment by kevaundray** [comment]
As of this commit, the native non-wasm version is passing tests. The tests in Test-ubuntu that are failing are related to `compiler::wasm`.

File: compiler/noirc_driver/src/stdlib.rs
Comment (kevaundray) [inline compiler/noirc_driver/src/stdlib.rs:5]: Need to double check if this windows shenanigans is still needed -- I removed most of the windows specific config code that was in file_reader

File: compiler/wasm/src/compile.rs
Comment (kevaundray) [inline compiler/wasm/src/compile.rs:136]: Can bikeshed this name

Comment (kevaundray) [reply compiler/wasm/src/compile.rs:136]: We're loading everything in memory either way, so this is fine -- If we want to optimize I think first removing code from memory,  once its compiled is the lowest hanging fruit

File: compiler/wasm/src/compile.rs
Comment (kevaundray) [inline compiler/wasm/src/compile.rs:164]: We could make this a reference, so that Js code can pass multiple instances of this. The issue is that there needs to be a clone somewhere, so it may be better to make it explicit on the Js side

File: compiler/fm/src/lib.rs
Comment (kevaundray) [inline compiler/fm/src/lib.rs:50]:
```suggestion
        let file_name = self.root.join(file_name);
```

File: compiler/fm/src/lib.rs
Comment (kevaundray) [inline compiler/fm/src/lib.rs:54]: This and add_file_with_source are public, so we should add doc comments

File: compiler/fm/src/lib.rs
Comment (kevaundray) [inline compiler/fm/src/lib.rs:95]: I can create an issue for this -- its not related to this PR, so I left it as is

File: compiler/noirc_driver/src/lib.rs
Comment (kevaundray) [inline compiler/noirc_driver/src/lib.rs:94]: We could make this return the FileID and panic internally

File: compiler/wasm/src/compile.rs
Comment (kevaundray) [inline compiler/wasm/src/compile.rs:136]: We can do something similar to this with the DependencyGraph and pass it in as a parameter

**Comment by jfecher** [comment]
Is this PR intended for the LSP or for general cleanup? I thought it was related to wasm's filesystem oddities at first but it seems not.

**Comment by kevaundray** [comment]
> Is this PR intended for the LSP or for general cleanup? I thought it was related to wasm's filesystem oddities at first but it seems not.

Its intended for the odditities related to wasm's filesystem that has sort of spread into many components including the lsp -- This now allows WASM/typescript to fill in the filesystem, however they see fit (async/sync) using the SourceMap and then the rust compiler just consumes the source map and uses it to populate the file system.

There are some general cleanup items that arise out of it, for example, it is now the compiler that "knows" about the stdlib instead of having the FileManager know about the stdlib. Another is that we no longer need the SourceResolver crate

File: compiler/integration-tests/test/browser/compile_prove_verify.test.ts
Comment (kevaundray) [inline compiler/integration-tests/test/browser/compile_prove_verify.test.ts:39]: Can create issue to remove source resolver

File: compiler/wasm/src/compile.rs
Comment (kevaundray) [inline compiler/wasm/src/compile.rs:162]: Generally not a fan of constructs like Option<bool> -- I believe it is here so that we can supply a variable number of arguments in javascript land. Though because I put file_source_map as the last argument, it negates the usage 😅

File: compiler/fm/src/lib.rs
Comment (TomAFrench) [inline compiler/fm/src/lib.rs:116]: Not sure it should as it will just immediately create a pathbuf again. We should probably rename `from_path` to `from_path_buf` though.

Comment (kevaundray) [reply compiler/fm/src/lib.rs:116]: I think that's an issue with PathString that has leaked into `name_to_id` -- since I shouldn't need to allocate in order to use something as a key lookup. We could probably get rid of PathString and or just use a type alias, though I was going to leave that to another PR

File: compiler/wasm/src/compile.rs
Comment (TomAFrench) [reply compiler/wasm/src/compile.rs:164]: Agreed, no sense in forcing an unnecessary clone in the case where we only pass one instance.

File: compiler/wasm/test/node/index.test.ts
Comment (alexghr) [inline compiler/wasm/test/node/index.test.ts:58]: Hmm, would this work with `lib_a/src/lib.nr`?

@kevaundray I forgot to mention it in our call yesterday but Aztec's noir-compiler wrapper uses source-resolver to dynamically remap library files to their location on disk. E.g. compiler would ask for `lib_a/lib.nr` and we'd use source-resolver to remap that to `/some/directory/lib_a/src/lib.nr`.

I think thi

Comment (alexghr) [reply compiler/wasm/test/node/index.test.ts:58]: I think we'll have to remap the files when filling the `PathToFileSourceMap` so that the compiler finds the files where it expects them to be (e.g. map the file on disk `/home/alex/some_lib/src/lib.nr` to `some_lib/lib.nr`)

https://github.com/noir-lang/noir/blob/d120852473233488ba11aeb40f010b06f6cae735/compiler/wasm/src/compile.rs#L281-L284

Comment (TomAFrench) [reply compiler/wasm/test/node/index.test.ts:58]: I think that we should have the dependency name and path be separated and both passed in tbh.

Currently we don't have a good way to handle a package which is depended on by two other packages which use different aliases for it, this will remove the need for remapping the files at the same time.

Comment (alexghr) [reply compiler/wasm/test/node/index.test.ts:58]: Hmm, it should probably also handle aliasing libraries in case there's two crates depending on the same library (with the same name) but different versions. :thinking:

File: compiler/wasm/src/compile.rs
Comment (TomAFrench) [reply compiler/wasm/src/compile.rs:162]: Let's leave this as is for now and we can modify this in a followup.

File: compiler/noirc_driver/src/stdlib.rs
Comment (TomAFrench) [reply compiler/noirc_driver/src/stdlib.rs:5]: It should still be necessary, although we're not supporting Windows officially anymore it's not a particularly burdensome bit of code.

File: compiler/fm/src/lib.rs
Comment (TomAFrench) [reply compiler/fm/src/lib.rs:63]: I've added some doc comments. I think they hit the main gotchas for these methods.

File: compiler/wasm/test/node/index.test.ts
Comment (TomAFrench) [reply compiler/wasm/test/node/index.test.ts:58]: I don't think these changes create any new issues for noir-compiler though right? You'll just need to move the path modification to be removing the root directory when adding it to the `PathToFileSourceMap` rather than removing it in the source-resolver.

We can fix this so that the wasm compiler doesn't have this strange behaviour later though I think.

File: compiler/integration-tests/test/browser/compile_prove_verify.test.ts
Comment (TomAFrench) [reply compiler/integration-tests/test/browser/compile_prove_verify.test.ts:39]: Removed in https://github.com/noir-lang/noir/pull/3791

File: tooling/nargo/src/lib.rs
Comment (TomAFrench) [inline tooling/nargo/src/lib.rs:87]:
```suggestion
```

File: compiler/noirc_driver/src/lib.rs
Comment (TomAFrench) [reply compiler/noirc_driver/src/lib.rs:94]: I don't think a panic is necessarily the correct way to handle this. We should have an error for the compiler looking for a file which does not exist.

I'm fine leaving this as a panic here for now however.

**Comment by TomAFrench** [comment]
I think this doesn't cause any major problems beyond the expected need to modify code interacting with `noir_wasm`. I'm good to merge this once followup issues have been made.

**Comment by kevaundray** [comment]
> I think this doesn't cause any major problems beyond the expected need to modify code interacting with `noir_wasm`. I'm good to merge this once followup issues have been made.

@alexghr pinging you to confirm this is okay too?

**Comment by kevaundray** [comment]
I'll also quickly put together a rough PR for a different API which aligns more with what we have in nargo that we can merge in the future

**Comment by kevaundray** [comment]
(non-breaking) PR is here for a different API that is more context-centric https://github.com/noir-lang/noir/pull/3798

File: compiler/noirc_driver/src/lib.rs
Comment (kevaundray) [reply compiler/noirc_driver/src/lib.rs:102]: Alright, I'll also add an issue for this

File: compiler/fm/src/lib.rs
Comment (kevaundray) [reply compiler/fm/src/lib.rs:130]: https://github.com/noir-lang/noir/issues/3799

File: compiler/fm/src/lib.rs
Comment (kevaundray) [reply compiler/fm/src/lib.rs:104]: https://github.com/noir-lang/noir/issues/3799

**Comment by alexghr** [comment]
Yep, looks good to me! I think we just have to be aware of having to update noir-compiler in aztec-packages at the same time as the noir subrepo is updated in it.

**Review by alexghr — APPROVED** [review]

File: tooling/nargo/src/lib.rs
Comment (kevaundray) [reply tooling/nargo/src/lib.rs]: Modified here: https://github.com/noir-lang/noir/pull/3760/commits/3e26e3ca4f0820006ed729903a94d34486f993bc

**Review by TomAFrench — APPROVED** [review]

**Pull request merged**
